sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -321,6 +321,58 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     }
   }
 
+  private def updatePartitioningByAliases(exprs: Seq[NamedExpression], partioning: Partitioning)
+    : Partitioning = {
+    val aliasSeq = exprs.flatMap(_.collectFirst {
```

Comment:
this might do more than you'd like it to (at least if it behaves the way I understand collect first), i.e.
`df.select($"x" as "x1, struct($"a" as "a1", $"b" as "b1") as "s1")`

_x1_ and _s1_ are aliases, _a1_ and _b1_ are not. it could even get more complicated if there was an _a1_ alias in the top level projections list.

Reply:
This pr only focuses on aliases, so the point you described above is out-of-scope in this pr. IMO more complicated cases should be fixed in follow-ups.

Reply:
@maropu , I didn't aim for supporting complex partitioning expressions (which deserves its own separate PR), I meant that this code could introduce regressions by 'over-capturing' nested aliases.

* my specific example is wrong since struct is transformed into a named struct (alias is replaced by an explicit name).